torch_geometric/utils/_negative_sampling.py

```diff
@@ -289,99 +356,152 @@ def structured_negative_sampling_feasible(
     num_nodes = maybe_num_nodes(edge_index, num_nodes)
     max_num_neighbors = num_nodes
 
-    edge_index = coalesce(edge_index, num_nodes=num_nodes)
-
     if not contains_neg_self_loops:
         edge_index, _ = remove_self_loops(edge_index)
         max_num_neighbors -= 1  # Reduce number of valid neighbors
 
     deg = degree(edge_index[0], num_nodes)
-    # True if there exists no node that is connected to all other nodes.
-    return bool(torch.all(deg < max_num_neighbors))
+    # structured sample is feasible if, for each node, deg > max_neigh/2
+    return bool(torch.all(2 * deg <= max_num_neighbors))
```

Comment:
```suggestion
    # structured sample is feasible if, for each node, deg > max_neigh/2
    return bool(torch.all(2 * deg <= max_num_neighbors))
```
The assumption here is that we don't want the negative edges to repeat. I like the assumption, but it might be too tight, also we don't guarantee that the negative edges need to be unique. I'd suggest keeping the old condition.

Reply:
I understand your point, and this makes me wondering the following:

1. **Should the negative sampling return duplicate edges?**
In this new implementation, `structured_negative_sampling` with `method=dense` guarantees that the negative edges are unique. Instead, when `method=sparse`, this is not explicitly guaranteed (but it should be very unlikely especially if the graph is sparse). However, this could be easily resolved by removing the duplicates after the sampling. Also, it is worth highlighting that the `negative_sampling` **never** returns duplicates.

2. **What is the purpose of `structured_negative_sampling_feasible`?**
In my understanding, this function checks if a structured negative sampling is feasible in a graph. Note that this does not mean that the function `structured_negative_sampling` will return all the negative edges requested. In fact, when the sparse method is selected, the method can fail (i.e. it is not able to sample a negative edge for all the true edges) even if it is feasible due to the randomness of the sampling. Vice versa, when the dense method is used, the method will always return the correct set of negative edges (if it is feasible) since it enumerates all edges.